### ISSUE TITLE
seatsio.SeatsioException#fromJson: Illegal reflective access

### ISSUE THREAD

**Issue by bryan-nz** [seatsio/seatsio-java#91]
With Java 17, **_seatsio.SeatsioException#fromJson_** will fail with the follow exception:

```
com.google.gson.JsonIOException: Failed making field 'java.lang.Throwable#detailMessage' accessible; either increase its visibility or write a custom TypeAdapter for its declaring type.
        at com.google.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:38)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:287)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:130)
        at com.google.gson.Gson.getAdapter(Gson.java:546)
        at com.google.gson.Gson.fromJson(Gson.java:1213)
        at com.google.gson.Gson.fromJson(Gson.java:1124)
        at com.google.gson.Gson.fromJson(Gson.java:1034)
        at com.google.gson.Gson.fromJson(Gson.java:969)
        at seatsio.SeatsioException.fromJson(SeatsioException.java:59)
        at seatsio.SeatsioException.from(SeatsioException.java:38)
        ...
```

With Java 11 (and presumably >= 9), it will succeed, but the following message will be written to stderr:
```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.gson.internal.reflect.ReflectionHelper to field java.lang.Throwable.detailMessage
WARNING: Please consider reporting this to the maintainers of com.google.gson.internal.reflect.ReflectionHelper
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

I believe the issue is because **_SeatsioException_** ultimately derives from Throwable, which has many private fields. A custom TypeAdapter will be required.

**Comment by mroloux** [comment]
Hi!

Which JDK are you on?

I tried reproducing this issue on Liberica JDK 17, but no luck so far.

**Comment by bryan-nz** [comment]
OpenJDK:
```
openjdk version "17.0.4" 2022-07-19
OpenJDK Runtime Environment (build 17.0.4+8-Debian-1deb11u1)
OpenJDK 64-Bit Server VM (build 17.0.4+8-Debian-1deb11u1, mixed mode, sharing)
```

As you've discovered, it seems that it doesn't fail under unit testing! But it does fail at runtime, and I'm not sure why.

This will succeed:
```java
    @Test
    public void testJson() {
        assertThrows(SeatsioException.class, () -> {
            throw SeatsioException.fromJson("{\"status\":404,\"messages\":[\"Event not found: 9bf37326-554a-419b-8859-394b6a526374\"],\"errors\":[{\"code\":\"EVENT_NOT_FOUND\",\"message\":\"Event not found: 9bf37326-554a-419b-8859-394b6a526374\"}],\"warnings\":[],\"requestId\":\"603276b8-9de5-42ad-8581-6afa3044399a\"}".getBytes());
        });
    }
```

But the equivalent application will fail:
```java
public class TestApp {
    public static void main(String[] args) {
        SeatsioException.fromJson("{\"status\":404,\"messages\":[\"Event not found: INVALID\"],\"errors\":[{\"code\":\"EVENT_NOT_FOUND\",\"message\":\"Event not found: INVALID\"}],\"warnings\":[],\"requestId\":\"a0ce38df-556f-483e-b50b-08e4214ab350\"}".getBytes());
    }
}
```
```
Exception in thread "main" com.google.gson.JsonIOException: Failed making field 'java.lang.Throwable#detailMessage' accessible; either increase its visibility or write a custom TypeAdapter for its declaring type.
	at com.google.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:38)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:287)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:130)
	at com.google.gson.Gson.getAdapter(Gson.java:546)
	at com.google.gson.Gson.fromJson(Gson.java:1213)
	at com.google.gson.Gson.fromJson(Gson.java:1124)
	at com.google.gson.Gson.fromJson(Gson.java:1034)
	at com.google.gson.Gson.fromJson(Gson.java:969)
	at seatsio.SeatsioException.fromJson(SeatsioException.java:59)
	at seatsio.TestApp.main(TestApp.java:5)
```

**Comment by mroloux** [comment]
Good catch. Gradle seems to add options to the JVM to work around this. If you run code without Gradle, the errors appear obviously.

Should be fixed now (v73.1.0).

**Comment by mroloux** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Block direct calls to seatsio-core-v2.herokuapp.com](https://trello.com/c/AHpECuRs/6092-block-direct-calls-to-seatsio-core-v2herokuappcom)
